lib/Target/AVR/InstPrinter/AVRInstPrinter.cpp

```diff
@@ -76,15 +78,24 @@ void AVRInstPrinter::printInst(const MCInst *MI, raw_ostream &O,
 }
 
 const char *
-AVRInstPrinter::getPrettyRegisterName(unsigned RegNo, MCRegisterInfo const& MRI) {
+AVRInstPrinter::getPrettyRegisterName(unsigned RegNum, MCRegisterInfo const& MRI) {
+  unsigned RegToPrint = RegNum;
+
 #ifdef LLVM_AVR_GCC_COMPAT
-  if (AVRMCRegisterClasses[AVR::DREGSRegClassID].contains(RegNo) &&
-      !AVRMCRegisterClasses[AVR::PTRREGSRegClassID].contains(RegNo))
+  // GCC prints register pairs by just printing the lower register
+  // If the register contains a subregister, print it instead
+  if (MRI.getSubReg(RegNum, 1) != 0)
   {
-    return getRegisterName(MRI.getSubReg(RegNo, AVR::sub_lo));
+    auto RegLoNum = MRI.getSubReg(RegNum, AVR::sub_lo);
+
+    if(RegLoNum == AVR::NoRegister)
+      RegToPrint = RegNum; // print the entire register pair
+    else
+      RegToPrint = RegLoNum; // print the lower register
   }
```

Comment:
The whole if block is odd... Let's rewrite the condition without magic numbers. We get:

````
if (MRI.getSubReg(RegNum, AVR::sub_lo) != AVR::NoRegister)
````

Now it's easy to see that it checks the same thing twice. Let's rewrite the whole thing as:

````
  auto RegLoNum = MRI.getSubReg(RegNum, AVR::sub_lo);
  return getRegisterName(RegLoNum != AVR::NoRegister ? RegLoNum : RegNum);
````

... or some such.